src/snowflake/snowpark/_internal/data_source/jdbc.py

```diff
@@ -0,0 +1,496 @@
+    "java.lang.Integer": DecimalType,
+    "java.lang.Short": DecimalType,
+}
+JDBC_TYPE_TO_SNOWFLAKE_TYPE = {
+    "NUMERIC": DecimalType,
+    "VARCHAR": StringType,
+    "CHAR": StringType,
+    "CLOB": StringType,
+    "NCHAR": StringType,
+    "NVARCHAR": StringType,
+    "NCLOB": StringType,
+    "TIMESTAMP": TimestampType,
+    "VARBINARY": StringType,
+}
+
+PARTITION_TABLE_COLUMN_NAME = "partition"
+
+
+class JDBC:
+    def __init__(
+        self,
+        session: "Session",
+        url: str,
+        table_or_query: str,
+        external_access_integration: str,
+        imports: List[str],
+        is_query: bool,
+        secret: str,
+        *,
+        properties: Optional[dict] = None,
+        packages: Optional[List[str]] = None,
+        java_version: Optional[int] = 11,
+        column: Optional[str] = None,
+        lower_bound: Optional[Union[str, int]] = None,
+        upper_bound: Optional[Union[str, int]] = None,
+        num_partitions: Optional[int] = None,
+        query_timeout: Optional[int] = 0,
+        fetch_size: Optional[int] = 0,
+        custom_schema: Optional[Union[str, StructType]] = None,
+        predicates: Optional[List[str]] = None,
+        session_init_statement: Optional[List[str]] = None,
+        _emit_ast: bool = True,
+    ) -> None:
+        self.session = session
+        self.url = url
+        self.table_or_query = table_or_query
+
+        self.external_access_integration = external_access_integration
+        self.imports = imports
+        self.packages = packages
+        self.java_version = java_version
+        self.secret = secret
+        self.properties = properties
+
+        self.is_query = is_query
+        self.column = column
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.num_partitions = num_partitions
+        self.query_timeout = query_timeout
+        self.fetch_size = fetch_size
+        self.custom_schema = custom_schema
+        self.predicates = predicates
+        self.session_init_statement = session_init_statement
+        self.raw_schema = None
+        self._emit_ast = _emit_ast
+        self.imports_sql = (
+            f"""IMPORTS =({",".join([f"'{imp}'" for imp in self.imports])})"""
+            if self.imports is not None
+            else ""
+        )
+        self.packages_sql = (
+            f"""PACKAGES=({','.join([f"'{pack}'" for pack in self.packages])})"""
+            if self.packages is not None
+            else "PACKAGES=('com.snowflake:snowpark:latest')"
+        )
+        self.secret_sql = f"SECRETS = ('cred' = {self.secret})"
+        self._infer_schema_successful = True
+        self.secret_detector()
+
+    @cached_property
+    def schema(self) -> StructType:
+        infer_schema_udtf_name = random_name_for_temp_object(TempObjectType.FUNCTION)
+        infer_schema_udtf_registration = f"""
+            CREATE OR REPLACE FUNCTION {infer_schema_udtf_name}(query VARCHAR)
```

Comment:
is it creating a permanent function? do we need a TEMPORARY keyword?
```suggestion
            CREATE OR REPLACE TEMPORARY FUNCTION {infer_schema_udtf_name}(query VARCHAR)
```